crates/optimism/flashblocks/src/consensus.rs

```diff
@@ -0,0 +1,81 @@
+use crate::FlashBlockCompleteSequenceRx;
+use alloy_primitives::B256;
+use reth_node_api::{ConsensusEngineHandle, EngineApiMessageVersion};
+use reth_optimism_payload_builder::OpPayloadTypes;
+use ringbuffer::{AllocRingBuffer, RingBuffer};
+use tracing::warn;
+
+/// Consensus client that sends FCUs and new payloads using blocks from a [`FlashBlockService`]
+///
+/// [`FlashBlockService`]: crate::FlashBlockService
+#[derive(Debug)]
+pub struct FlashBlockConsensusClient {
+    /// Handle to execution client.
+    engine_handle: ConsensusEngineHandle<OpPayloadTypes>,
+    sequence_receiver: FlashBlockCompleteSequenceRx,
+}
+
+impl FlashBlockConsensusClient {
+    /// Create a new `FlashBlockConsensusClient` with the given Op engine and sequence receiver.
+    pub const fn new(
+        engine_handle: ConsensusEngineHandle<OpPayloadTypes>,
+        sequence_receiver: FlashBlockCompleteSequenceRx,
+    ) -> eyre::Result<Self> {
+        Ok(Self { engine_handle, sequence_receiver })
+    }
+
+    /// Get previous block hash using previous block hash buffer. If it isn't available (buffer
+    /// started more recently than `offset`), return default zero hash
+    fn get_previous_block_hash(
+        &self,
+        previous_block_hashes: &AllocRingBuffer<B256>,
+        offset: usize,
+    ) -> B256 {
+        *previous_block_hashes
+            .len()
+            .checked_sub(offset)
+            .and_then(|index| previous_block_hashes.get(index))
+            .unwrap_or_default()
+    }
+
+    /// Spawn the client to start sending FCUs and new payloads by periodically fetching recent
+    /// blocks.
+    pub async fn run(mut self) {
+        let mut previous_block_hashes = AllocRingBuffer::new(64);
+
+        loop {
+            match self.sequence_receiver.recv().await {
+                Ok(sequence) => {
+                    let block_hash = sequence.last().diff.block_hash;
+                    previous_block_hashes.push(block_hash);
```

Comment:
we cant use this

these hashes are all incorrect because state root isnt calculated

the need to use base.parent_hash

Reply:
yes for now
we can fix this in a followup or multiple ones